plugin-server/tests/shared/process-event.ts

```diff
@@ -2162,5 +2166,128 @@ export const createProcessEventTests = (
         })
     }
 
+    test('set and set_once on the same key', async () => {
+        await createPerson(hub, team, ['distinct_id1'])
+
+        await processEvent(
+            'distinct_id1',
+            '',
+            '',
+            {
+                event: 'some_event',
+                properties: {
+                    token: team.api_token,
+                    distinct_id: 'distinct_id1',
+                    $set: { a_prop: 'test-set' },
+                    $set_once: { a_prop: 'test-set_once' },
+                },
+            } as any as PluginEvent,
+            team.id,
+            now,
+            now,
+            new UUIDT().toString()
+        )
+        expect((await hub.db.fetchEvents()).length).toBe(1)
+
+        const [event] = await hub.db.fetchEvents()
+        expect(event.properties['$set']).toEqual({ a_prop: 'test-set' })
+        expect(event.properties['$set_once']).toEqual({ a_prop: 'test-set_once' })
+
+        const [person] = await hub.db.fetchPersons()
+        expect(await hub.db.fetchDistinctIdValues(person)).toEqual(['distinct_id1'])
+        expect(person.properties).toEqual({ a_prop: 'test-set' })
+    })
+
+    describe('ingestion in any order', () => {
+        const ts0: DateTime = now
+        const ts1: DateTime = now.plus({ minutes: 1 })
+        const ts2: DateTime = now.plus({ minutes: 2 })
+        const ts3: DateTime = now.plus({ minutes: 3 })
+        const set0: Properties = { s0123o0123: 's0a', s02o13: 's0b', s013: 's0e' }
```

Comment:
nit: these prop names follow a pattern to you but look like gibberish to other readers for a while before they really get it